index.bs

```diff
@@ -1350,69 +1858,166 @@ dictionary MLClampOptions {
 };
 </script>
-<div algorithm=concat>
+<div>
     **Arguments:**
         - *inputs*: a sequence of {{MLOperand}}. All input tensors must have the
             same shape, except for the size of the dimension to concatenate on.
-        - *axis*: an {{unsigned long}} scalar. The axis that the inputs concatenate along. Its value must be in the range [0, N-1] where N is the rank of input tensors.
+        - *axis*: an {{unsigned long}} scalar. The axis that the inputs concatenate along. Its value must be in the range [0, N-1] where N is the [=rank=] of the input tensors.
 
     **Returns:** an {{MLOperand}}. The concatenated tensor of all the inputs along
     the *axis*. The output tensor has the same shape except on the dimension
     that all the inputs concatenated along. The size of that dimension is
     computed as the sum of all the input sizes of the same dimension.
 </div>
 
-### The conv2d() method ### {#api-mlgraphbuilder-conv2d}
-Compute a 2-D convolution given 4-D input and filter tensors
-<script type=idl>
-enum MLConv2dFilterOperandLayout {
+<details open algorithm>
+
+  <summary>
+    The <dfn method for=MLGraphBuilder>concat(|inputs|, |axis|)</dfn> method steps are:
+  </summary>
+  <div class=algorithm-steps>
+    <div class="note">
+        The permissions and context validity have been checked by [[#api-mlgraphbuilder-constructor]] steps.
+    </div>
+    1. [=Assert=]: the type of |inputs| is sequence of {{MLOperand}} objects.
+    1. [=Assert=]: the type of |axis| is `unsigned long`.
+    1. [=Assert=]: the shape, i.e. {{MLOperandDescriptor/dimensions}} of each operand in |inputs| is the same, except on the dimension given by |axis| on which they are concatenated.
+    1. [=Assert=]: the {{MLOperandDescriptor/type}} of each operand in |inputs| is the same.
+    1. If any of the following steps fail, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+        1. Let |desc| be |inputs|[0].{{MLOperand/[[descriptor]]}}.
+        1. If |axis| is greater than or equal to the [=rank=] of |desc|, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+        1. Let |desc|.{{MLOperandDescriptor/dimensions}}[|axis|] be 0.
+        1. [=map/For each=] |index| in [=the range=] 0 to the [=rank=] of |inputs|, exclusive:
+            1. If <a>validating MLOperand</a> given |inputs|[|index|] and [=this=] returns false, then [=exception/throw=] a "{{DataError}}" {{DOMException}}.
+            1. [=map/For each=] |dim| in [=the range=] 0 to the [=rank=] of |inputs|[|index|].{{MLOperandDescriptor/dimensions}}, exclusive:
+                <div class="note">
+                    If the shape of each corresponding dimension and type of the operands, except for those of the dimension given by |axis|, is not the same, fail.
+                </div>
+                1. If |inputs|[|index|].{{MLOperandDescriptor/type}} is not equal to |inputs|[0].{{MLOperandDescriptor/type}}.
```

Comment:
Behind L1997, they are validating the dimensions. Validating for each input right behind the L1995 may be better.

Reply:
OK, I got it :) will commit soon.